### ISSUE TITLE
Features disappearing when zoom changes

### ISSUE THREAD

**Issue by Jesus89** [CartoDB/carto-vl#91]
For some values of the zoom this feature is not rendered.

(In this GIF I am zooming out the map)

![bug-render-zoom](https://user-images.githubusercontent.com/2227572/36421017-3d7b4516-1637-11e8-8848-1d5a9af567c6.gif)

**Comment by davidmanzanares** [comment]
How can this be reproduced?

**Comment by Jesus89** [comment]
Try [this example](https://cartodb.github.io/renderer-prototype/example/mapbox.html#eyJhIjoicG9pbnQiLCJiIjoiIiwiYyI6ImFycm95by1jYXJ0byIsImQiOiJodHRwczovL3t1c2VyfS5jYXJ0by5jb20iLCJlIjoid2lkdGg6IDIwMFxuY29sb3I6IHJnYmEoMCwxLDAsMC41KSIsImYiOnsibG5nIjotNi44OTU5MjY3NzI4ODM4NTQsImxhdCI6MjQuODAzMDU4MDY4NTQ3MTczfSwiZyI6MS42NTk5NjEzMzUyMjg3MTE3fQ==).

**Comment by davidmanzanares** [comment]
Seems like a Maps API issue after talking with @jgoizueta 

We saw that Maps API returns 'No content' on zoom levels greater than 4. This is probably due to the point being the (0,0) point.

![mapsapi](https://user-images.githubusercontent.com/1692175/36593817-a03153c4-189b-11e8-9ef6-61a438caf2ea.png)

**Comment by jgoizueta** [comment]
I took a look at this problem.

### TL;DR;
The disappearing point is not easy to fix, the problem is in node-mapnik.

The point-duplication has a different cause, the culprit it our aggregation and can be solved by changing the spatial filtering done in the aggregation:
```
   WHERE the_geom_webmercator && _cdb_params.bbox
```
to:
```
        WHERE ST_X(the_geom_webmercator) >= ST_XMIN(bbox)
          AND ST_X(the_geom_webmercator) < ST_XMAX(bbox)
          AND ST_Y(the_geom_webmercator) >= ST_YMIN(bbox)
          AND ST_Y(the_geom_webmercator) < ST_YMAX(bbox)
```
But this could degrade the performance (or not, because the `the_geom_webmercator && _cdb_params.bbox` is added by Mapnik anyway in its query wrapping).
A solution could be add both conditions (we'd need to check).

### The gory details

The duplication problem is that points between tiles (at tile borders) are selected and aggregate in multiple tiles by our `&&` filtering.

For example point at 0,0 (as in the example) belongs to 4 tiles at any zoom other than 0.

So why is the point duplicated in the example, rather than *quadruplicated*?

Because when transformed to WM it is moved slightly south, so it is between two tiles only:
```
Select st_astext(st_transform('SRID=4326;POINT(0 0)'::geometry,3857));
           st_astext
--------------------------------
 POINT(0 -7.08115455161362e-10)
```
(hey @Algunenano, maybe PROJ 5 will do this better?)

You can verify this by setting the WM coords to exactly 0,0:

```
update point set the_geom_webmercator='SRID=3857;POINT(0 0)'::geometry where cartodb_id=5;
```

You'll get 5 points in the map at low zooms instead of 3.

But wait, the point disappears at zoom>=5 !!

This is really puzzling, because the aggregation queries are not the culprit (they never miss any points).

I believe that it is being filtered by Mapnik Vector tiles (perhaps somewhere around [here](https://github.com/CartoDB/node-mapnik/blob/v3.6.2-carto/src/mapnik_cairo_surface.cpp))
because the bbox that Mapnik computes is also slightly off, but differently than
the PostGIS WM error: the Y coordinate becomes -6.984919309616089e-10 for the zoom 5 tiles.

If you set the WM coordinates to exactly 0 as above the missing point problem dissapears.

**Comment by dgaubert** [comment]
Is this happening for our Postgis MVT renderer (ST_AsMVT)?

**Comment by Algunenano** [comment]
> (hey @Algunenano, maybe PROJ 5 will do this better?)

```c
#include <proj_api.h>
#include <stdio.h>


typedef struct
{
	double x, y, z;
}
POINT;

int main()
{
    projPJ pj4326 = pj_init_plus("+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs +towgs84=0,0,0");
    projPJ pj3857 = pj_init_plus("+proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0 +k=1.0 +units=m +nadgrids=@null +wktext  +no_defs");

    POINT p1 = { 0.0, 0.0, 0.0 };
    pj_transform(pj4326, pj3857, 1, 0, &(p1.x), &(p1.y), &(p1.z));
    printf("%.12f %.12f %.12f\n", p1.x, p1.y, p1.z);
}
```

```bash
$ ./a.out 
0.000000000000 -0.000000000708 0.000000000000
```

So the deviation is still there in proj5.

**Comment by jgoizueta** [comment]
> Is this happening for our Postgis MVT renderer (ST_AsMVT)?

Good question; we should find out that, and also if ST_AsMVT supports date/time types, which is another problem we have now with Mapnik MVTs.

**Comment by jgoizueta** [comment]
The problem of points between tiles disappearing seems to be fixed in production now, but I don't know what fixed it. There are changes in PROJ 5.1 that might have an influence, but we don't have PROJ 5 in our production PostGIS.

@Algunenano could this (or any other change of yours) have fixed this? https://github.com/CartoDB/mapnik-vector-tile/commit/34157acce0c2a7c37cf0c79b4aa3326921d4bbe2

**Comment by Algunenano** [comment]
My change was a fix to make sure variables with equivalent value (0, FALSE, 0.0) but different type were returned together, so that shouldn't be the cause; but as part of the fork I also updated the module from 1.5.0 to 1.6.1 which includes the commit you mention: `Solved problem surrounded odd reprojection issues when dealing with low zoom level tiles and projections covering small geographic areas.`

It could also have been fixed when we removed the extra simplification that was being done to the geometries (https://github.com/CartoDB/tilelive-bridge/commit/0fefba9d623175d8e57d42592b62fc93aedff663).

Certainly PROJ 5.1 wasn't the cause

If you need to be sure just tell me how to reproduce this locally and I'll investigate further.

**Comment by jgoizueta** [comment]
Since this doesn't seem a problem anymore I'm closing this. The related point-between tiles duplication is being handled in #541

**Comment by jgoizueta** [comment]
This problem went away when we were using a nonzero buffersize. I has reappeared when we changed the buffersize back to 0.

Se #541 : point features *between tiles* (with coordinates on the border) sometimes are not present in the mapnik tiles. For example the (0,0) point at zoom level 2 should appear in tiles 1,1, 2,1, 1,2 and 2,2. But with a zero buffer it dissappears from the northern tiles 1,1, 2,1. So this particular case is probably caused by the PROJ.4 webmercator accuracy that moves 0,0, slightly to the South. (this problem is fixed in v5.1, but PostGIS is not using that yet)

**Comment by rochoa** [comment]
This seems to be working OK. 😄 

Feel free to re-open if the issue persists.
